File: reference/src/security/interfaces.ts
/**
 * Enterprise security interfaces for AI-SDLC Framework.
 * Defines contracts for sandbox isolation, JIT credentials,
 * kill switches, and approval workflows.
 */

// ── Sandbox Isolation ────────────────────────────────────────────────

export type NetworkPolicy = 'none' | 'egress-only' | 'full';

export interface SandboxConstraints {
  maxMemoryMb: number;
  maxCpuPercent: number;
  networkPolicy: NetworkPolicy;
  timeoutMs: number;
  allowedPaths: string[];
}

export type SandboxStatus = 'idle' | 'running' | 'terminated' | 'error';

export interface Sandbox {
  /** Isolate an agent task with the given constraints. Returns a sandbox ID. */
  isolate(taskId: string, constraints: SandboxConstraints): Promise<string>;
  /** Destroy a running sandbox. */
  destroy(sandboxId: string): Promise<void>;
  /** Get the current status of a sandbox. */
  getStatus(sandboxId: string): Promise<SandboxStatus>;
}

// ── Secret Store ─────────────────────────────────────────────────────

export interface SecretStore {
  /** Resolve a secret by name. Returns undefined if not found. */
  get(name: string): string | undefined;
  /** Resolve a secret by name. Throws if not found. */
  getRequired(name: string): string;
  /** Store a secret (optional — read-only stores may omit). */
  set?(name: string, value: string, ttl?: number): Promise<void>;
  /** Delete a secret (optional — read-only stores may omit). */
  delete?(name: string): Promise<void>;
}

// ── JIT Credential Issuing ───────────────────────────────────────────

export interface JITCredential {
  id: string;
  token: string;
  scope: string[];
  issuedAt: string;
  expiresAt: string;
}

export interface JITCredentialIssuer {
  /** Issue a short-lived credential with the given scope and TTL. */
  issue(agentId: string, scope: string[], ttlMs: number): Promise<JITCredential>;
  /** Revoke an issued credential. */
  revoke(credentialId: string): Promise<void>;
  /** Check if a credential is still valid. */
  isValid(credentialId: string): Promise<boolean>;
}

// ── Kill Switch ──────────────────────────────────────────────────────

export interface KillSwitch {
  /** Activate the kill switch, halting all agent operations. */
  activate(reason: string): Promise<void>;
  /** Deactivate the kill switch, resuming operations. */
  deactivate(): Promise<void>;
  /** Check if the kill switch is currently active. */
  isActive(): Promise<boolean>;
  /** Get the reason the kill switch was activated, if active. */
  getReason(): Promise<string | undefined>;
}

// ── Approval Workflows ───────────────────────────────────────────────

export type ApprovalTier = 'auto' | 'peer-review' | 'team-lead' | 'security-review';

export type ApprovalStatus = 'pending' | 'approved' | 'rejected' | 'expired';

export interface ApprovalRequest {
  id: string;
  tier: ApprovalTier;
  requester: string;
  description: string;
  status: ApprovalStatus;
  createdAt: string;
  decidedAt?: string;
  decidedBy?: string;
}

export interface ApprovalWorkflow {
  /** Submit an action for approval. */
  submit(tier: ApprovalTier, requester: string, description: string): Promise<ApprovalRequest>;
  /** Approve a pending request. */
  approve(requestId: string, approver: string): Promise<ApprovalRequest>;
  /** Reject a pending request. */
  reject(requestId: string, rejector: string, reason: string): Promise<ApprovalRequest>;
  /** Get the status of an approval request. */
  getStatus(requestId: string): Promise<ApprovalRequest>;
}

File: sdk-go/security/interfaces.go
// Package security provides sandbox, secret management, JIT credentials, kill switch, and approval workflows.
package security

import "context"

// SandboxResult is the outcome of running a command in a sandbox.
type SandboxResult struct {
	ExitCode int    `json:"exitCode"`
	Stdout   string `json:"stdout"`
	Stderr   string `json:"stderr"`
}

// Sandbox defines an isolated execution environment.
type Sandbox interface {
	Execute(ctx context.Context, command string, args []string) (*SandboxResult, error)
	Cleanup(ctx context.Context) error
}

// SecretStore provides access to secrets.
type SecretStore interface {
	GetSecret(ctx context.Context, name string) (string, error)
	SetSecret(ctx context.Context, name, value string) error
	DeleteSecret(ctx context.Context, name string) error
	ListSecrets(ctx context.Context) ([]string, error)
}

// JITCredential represents a just-in-time credential.
type JITCredential struct {
	Name      string `json:"name"`
	Value     string `json:"value"`
	ExpiresAt string `json:"expiresAt"`
}

// JITCredentialIssuer issues and revokes short-lived credentials.
type JITCredentialIssuer interface {
	Issue(ctx context.Context, scope []string, ttl string) (*JITCredential, error)
	Revoke(ctx context.Context, credential *JITCredential) error
}

// KillSwitch provides emergency stop capability.
type KillSwitch interface {
	Activate(ctx context.Context, reason string) error
	Deactivate(ctx context.Context) error
	IsActive(ctx context.Context) (bool, error)
}

// ApprovalTier represents the required approval level.
type ApprovalTier string

const (
	TierAuto           ApprovalTier = "auto"
	TierPeerReview     ApprovalTier = "peer-review"
	TierTeamLead       ApprovalTier = "team-lead"
	TierSecurityReview ApprovalTier = "security-review"
)

// ApprovalRequest represents a request for approval.
type ApprovalRequest struct {
	ID           string       `json:"id"`
	Stage        string       `json:"stage"`
	Tier         ApprovalTier `json:"tier"`
	Requester    string       `json:"requester"`
	Description  string       `json:"description"`
	RequestedAt  string       `json:"requestedAt"`
}

// ApprovalResponse represents the response to an approval request.
type ApprovalResponse struct {
	RequestID string `json:"requestId"`
	Approved  bool   `json:"approved"`
	Approver  string `json:"approver"`
	Comment   string `json:"comment,omitempty"`
	Timestamp string `json:"timestamp"`
}

// ApprovalWorkflow manages approval requests.
type ApprovalWorkflow interface {
	Request(ctx context.Context, req *ApprovalRequest) error
	Respond(ctx context.Context, resp *ApprovalResponse) error
	GetStatus(ctx context.Context, requestID string) (*ApprovalRequest, *ApprovalResponse, error)
}
